### PR TITLE
Support data loading for GSVA data.

### PR DESCRIPTION
**Pull request by pieterlukasse**
# What? Why?

Summary: Updated import code for supporting data loading for GSVA data. 

**Requires updated database schema in #2112**

Documentation in [docs/Import-Gene-Sets.md](https://github.com/thehyve/cbioportal/blob/fbf095c5153c301ec041f0f79bb41dc38d4e95fb/docs/Import-Gene-Sets.md)

**Changes by** @angelicaochoa and @sandertan 

* Made changes to support new `GeneticAlterationType` for geneset data in `GeneticAlterationType.java`
* Made changes to `DaoGeneSet.java` to support adding, deleting, updating geneset records
* Updated `ImportTabDelimData.java` to support import of GSVA data
* Updated `ResetDatabase.java` to also delete records from geneset tables
* Added model for geneset records (`GeneSet.java`)

Added features:

- Importer for gene set data
- Importer for supplementary gene set data
- Importer for gsva score and gsva pvalue data
- Importer for gene set hierarchy data
- Validator for gsva score and gsva pvalue data
- Unit test for import gene set data
- Unit test for import gene set hierarchy data
- Unit test for validate gsva score and gsva pvalue data
- Add GENESET_SCORE test studies for GSVA and P-VALUE
- Add gene set database as replacement for portal info for validator
- Add parser for hierarchy in yaml format
- Add documentation
- Rename old GeneSet class to SetOfGenes
- Move Geneset.java from model/ to core/..../portal/model/

#### Testing
Also added examples to test loading data; 

# Notify reviewers
@angelicaochoa @n1zea144

### CODE REVIEW

**Comment by pieterlukasse** [comment]
@sandertan could you update the description of this PR to reflect the actual contents?

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoGeneset.java [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+     * @return Geneset record
+     * @throws SQLException
+     * @throws DaoException 
+     */
+    private static Geneset extractGeneset(ResultSet rs) throws SQLException, DaoException {
+        Integer id = rs.getInt("ID");
+        Integer geneticEntityId = rs.getInt("GENETIC_ENTITY_ID");
+        String externalId = rs.getString("EXTERNAL_ID");
+        String name = rs.getString("NAME");
+        String description = rs.getString("DESCRIPTION");
```
Comment (angelicaochoa) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoGeneset.java]: The version of the `geneset` table that I have in my own test database contains the following columns:

```
+-------------------+--------------+------+-----+---------+-------+
| Field             | Type         | Null | Key | Default | Extra |
+-------------------+--------------+------+-----+---------+-------+
| ID                | int(11)      | NO   | PRI | NULL    |       |
| GENETIC_ENTITY_ID | int(11)      | NO   | UNI | NULL    |       |
| EXTERNAL_ID       | varchar(100) | NO   | UNI | NULL    |       |
| NAME_SHORT        | varchar(100) | NO   | UNI | NULL    |       |
| NAME              | varchar(300) | NO   |     | NULL    |       |
| REF_LINK          | text         | YES  |     | NULL    |       |
| VERSION           | varchar(50)  | NO   |     | NULL    |       |
+-------------------+--------------+------+-----+---------+-------+
```

There's no `DESCRIPTION` column in my version of this table... whatever version of these tables that I have are from back in December when I first started working with @pieterlukasse on supporting GSVA data loading so the result set can't be parsed correctly here. Was this column added in a subsequent version of these geneset-related tables? Where can we find the changes to `cgds.sql` and `migration.sql`? @pieterlukasse @sandertan

File: core/src/main/java/org/mskcc/cbio/portal/scripts/ResetDatabase.java [hunk 2]
```diff
@@ -59,9 +59,10 @@ public static void resetAnySizeDatabase() throws DaoException {
         DaoCancerStudy.deleteAllRecords();
         DaoGeneOptimized daoGene = DaoGeneOptimized.getInstance();
         daoGene.deleteAllRecords();
+        DaoGeneset.deleteAllRecords();
```
Comment (sandertan) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/ResetDatabase.java]: Yes indeed the `DaoGenesetInfo` should also be reset here. Will fix it!

The `DaoGenesetHierarchy` is actually reset in `DaoGeneset.deleteAllRecords()`, and the `DaoGenesetHierarchyLeaf` records are removed when records from `DaoGenesetHierarchy` are removed. There might be a better way to make this clear. I'll discuss it with @pieterlukasse

**Comment by sandertan** [comment]
@angelicaochoa thank you for your comments! Indeed we have changed the database schema slightly, and put all these changes in a seperate PR: #2112 

We should have notified you that it's required to update the database to that version first. Sorry!

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoGenesetHierarchyLeaf.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+		Connection connection = null;
+        PreparedStatement preparedStatement = null;
+        ResultSet resultSet = null;
+        
+        try {
+        	// Open connection to database
+            connection = JdbcUtil.getDbConnection(DaoGenesetHierarchyLeaf.class);
+	        
+	        // Prepare SQL statement
+            preparedStatement = connection.prepareStatement("SELECT * FROM geneset_hierarchy_lef WHERE NODE_ID = ?");
```
Comment (sandertan) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoGenesetHierarchyLeaf.java]: Thx, fixed it.

File: core/src/main/java/org/mskcc/cbio/portal/Geneset.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
+/*
```
Comment (sandertan) [inline core/src/main/java/org/mskcc/cbio/portal/Geneset.java]: Yes you're right, will move them!

File: core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGenesetHierarchy.java [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+            if (!options.has(data)) {
+                throw new UsageException(
+                        progName, description, parser,
+                        "'data' argument required");
+            }
+            
+            
+            File genesetFile = new File(options.valueOf(data));
+            
+        	System.out.println("Input file:");
```
Comment (sandertan) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGenesetHierarchy.java]: Yes, thx will change it.

File: core/src/main/java/org/mskcc/cbio/portal/scripts/ImportProfileData.java [hunk 6]
```diff
@@ -88,9 +88,11 @@ public void run() {
             }
        }
        catch (RuntimeException e) {
```
Comment (sandertan) [inline core/src/main/java/org/mskcc/cbio/portal/scripts/ImportProfileData.java]: Thx, changed it.

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java [hunk 7]
```diff
@@ -516,6 +516,8 @@ static void deleteCancerStudy(int internalCancerStudyId) throws DaoException {
                 "delete from patient where CANCER_STUDY_ID=?;",
                 "delete from copy_number_seg where CANCER_STUDY_ID=?;",
                 "delete from sample_list where CANCER_STUDY_ID=?;",
+                "delete from genetic_profile_link where REFERRING_GENETIC_PROFILE_ID IN (select GENETIC_PROFILE_ID FROM genetic_profile where CANCER_STUDY_ID=?);",
```
Comment (sandertan) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java]: In #2112

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoGenesetHierarchy.java [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+		Connection connection = null;
+        PreparedStatement preparedStatement = null;
+        ResultSet resultSet = null;
+        
+        try {
+        	// Open connection to database
+            connection = JdbcUtil.getDbConnection(DaoGenesetHierarchy.class);
+	        
+	        // Prepare SQL statement
+            preparedStatement = connection.prepareStatement("SELECT * FROM geneset_hierarchy LIMIT 1");
```
Comment (sandertan) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoGenesetHierarchy.java]: Yes, but this is faster when table has many entries.

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoGeneset.java (hunk 1 above)
Comment (sandertan) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoGeneset.java]: In #2112

**Comment by angelicaochoa** [comment]
@sandertan @pieterlukasse everything looks good to me - just ran the tests with the updated geneset tables 👍

**Comment by pieterlukasse** [comment]
replaced by #2166

**Pull request closed without merging**